### PR TITLE
feat(eibc): Expand list-demand-orders query with more parameters

### LINKED ISSUE #849: Add filters to `list-demand-oders`
- [ ] rollapp id
- [ ] type 
- [ ] limit

### PR DESCRIPTION
**Pull request by zale144**
## Description

----

Closes #849 

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ]  Targeted PR against the correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to the GitHub issue with discussion and accepted design
- [ ]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----;

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

---;

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

## Summary by CodeRabbit

- **New Features**
  - Introduced demand order filters for type, rollapp ID, and limit to enhance querying capabilities.

- **Refactor**
  - Refactored query and filtering mechanisms to support new demand order filters, improving maintainability and flexibility.

- **Bug Fixes**
  - Updated error handling for invalid statuses in demand order queries.

- **Tests**
  - Enhanced test cases to validate new filtering options and ensure robust demand order functionality.

- **Chores**
  - Updated documentation and comments to reflect the new demand order filter features.

### CODE REVIEW

File: x/eibc/client/cli/query_command_orders.go [hunk 1]
```diff
@@ -1,30 +1,63 @@
...
+					return fmt.Errorf("invalid packet type: %s", args[2])
+				}
+				request.Type = commontypes.Type(ptype)
+			}
+			if len(args) > 3 {
+				limit, err := strconv.ParseInt(args[3], 10, 32)
+				if err != nil {
+					return fmt.Errorf("limit must be an integer: %s", args[3])
+				}
+				request.Limit = int32(limit)
```
Comment (trinitys7) [inline x/eibc/client/cli/query_command_orders.go:52]: Any reason why we use limit instead of Pagination in cosmos-sdk?

Comment (zale144) [reply x/eibc/client/cli/query_command_orders.go:52]: well, the intent of this added filtering is just for debugging, it's not supposed to be for end users

File: proto/dymension/eibc/demand_order.proto [hunk 2]
```diff
@@ -24,5 +25,7 @@ message DemandOrder {
     ];
     string recipient = 5;
     bool is_fullfilled = 6;
-    dymensionxyz.dymension.common.Status tracking_packet_status = 8;
+    string rollapp_id = 7;
+    common.Type type = 8;
```
Comment (mtsitrin) [inline proto/dymension/eibc/demand_order.proto:29]: why one is `dymensionxyz.dymension.common.` and the other `common.`

File: proto/dymension/eibc/query.proto [hunk 3]
```diff
@@ -43,7 +45,13 @@ message QueryGetDemandOrderRequest {
 // QueryDemandOrdersByStatusRequest is the request type for the Query/GetDemandOrdersByStatus RPC method.
 message QueryDemandOrdersByStatusRequest {
   // id of the demand order to get
```
Comment (mtsitrin) [inline proto/dymension/eibc/query.proto:47]: wrong comment

File: x/delayedack/types/rollapp_packets_list_filter.go [hunk 4]
```diff
@@ -26,6 +26,7 @@ func PendingByRollappIDByMaxHeight(
 				End:   commontypes.RollappPacketByStatusByRollappIDByProofHeightPrefix(rollappID, status, maxProofHeight+1), // inclusive end
 			},
 		},
+		FilterFunc: func(packet commontypes.RollappPacket) bool { return true },
```
Comment (mtsitrin) [inline x/delayedack/types/rollapp_packets_list_filter.go:29]: should define it (`func(packet commontypes.RollappPacket) bool { return true }` )and reuse for readability

**Review by mtsitrin — CHANGES_REQUESTED** [review]

File: proto/dymension/eibc/demand_order.proto
Comment (mtsitrin) [inline proto/dymension/eibc/demand_order.proto:1]: state breaking? will require migration

File: x/eibc/keeper/keeper.go [hunk 5]
```diff
@@ -177,6 +178,11 @@ func (k Keeper) ListDemandOrdersByStatus(ctx sdk.Context, status commontypes.Sta
 	for ; iterator.Valid(); iterator.Next() {
 		var val types.DemandOrder
 		k.cdc.MustUnmarshal(iterator.Value(), &val)
+		for _, opt := range opts {
```
Comment (mtsitrin) [inline x/eibc/keeper/keeper.go:181]: the `limit(limit int)` option is bit miss the point here no?
u'll still process and unmarshal the entire query
I think better design for limit should break this iterator

**Review by mtsitrin — APPROVED** [review]

**Pull request merged**